File: datastructures/java/assignments/assignment3_BSTDeleteRoot/BinaryTree.java
// Class to represent Tree node
class Node {
    int data;
    Node left, right;

    public Node(int item) {
        data = item;
        left = null;
        right = null;
    }
}

public class BinaryTree {

    Node root = null;

    // print the in order traversal of the tree
    public void inOrder(Node root) {
        if (root != null) {
            inOrder(root.left);
            System.out.print(root.data + " ");
            inOrder(root.right);
        }
    } // inOrder

    public Node deleteNode(Node root, int data) {
        if (root == null) {
            return root;
        }
        // if the number to delete is less than the current node, go left
        if (data < root.data) {
            root.left = deleteNode(root.left, data);
        }
        // if the number to delete is greater than the current node, go right
        else if (data > root.data) {
            root.right = deleteNode(root.right, data);
        }
        // if the number to delete is the current node
        else {
            // if either the left or right child is null, return that node
            if (root.left == null) {
                return root.right;
            } else if (root.right == null) {
                return root.left;
            }
            // replace the root with the leftmost leaf of the right tree
            root.data = leftmostLeaf(root.right);
            // delete the duplicate leaf node
            root.right = deleteNode(root.right, root.data);
        }
        return root;
    } // deleteNode

    // find the leftmost leaf of the tree
    public int leftmostLeaf(Node root) {
        int value = root.data;
        // keep going down the left children of the until a leaf is reached
        while (root.left != null) {
            value = root.left.data;
            root = root.left;
        }
        // return the value held by the leaf
        return value;
    } // leftmostLeaf

    public static void main(String[] args) {

        BinaryTree tree = new BinaryTree();
        tree.root = new Node(4);
        tree.root.left = new Node(2);
        tree.root.right = new Node(6);
        tree.root.left.left = new Node(1);
        tree.root.left.right = new Node(3);
        tree.root.right.left = new Node(5);
        tree.root.right.right = new Node(7);

        // print tree before deletion
        System.out.print("Tree before root deletion: ");
        tree.inOrder(tree.root);
        System.out.println();

        // delete root node and print tree again
        tree.deleteNode(tree.root, tree.root.data);
        System.out.print("Tree after root deletion: ");
        tree.inOrder(tree.root);

    } // main
}

File: data-structures/python/assignments/assignment3_BSTDeleteRoot/BinaryTree.py
class Node:
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data


def print_in_order(root):
    """Prints the in order traversal of tree."""
    if root:
        # recur on left child
        print_in_order(root.left)
        # print the data of node
        print(root.data, end=" ")
        # recur on right child
        print_in_order(root.right)


def leftmost_leaf(node: Node) -> int:
    """Find the leftmost leaf of the tree."""
    curr = node

    # loop down to find the leftmost leaf
    while curr.left is not None:
        curr = curr.left

    # return data of the leftmost leaf node
    return curr.data


def delete_node(root, data: int) -> Node:
    """Given a binary search tree and data to delete, this function
    deletes the data and returns the new root.
    """
    # base case
    if root is None:
        return root

    # if the data to be deleted is smaller than
    # the root's data then it lies in left subtree
    if data < root.data:
        root.left = delete_node(root.left, data)

    # if the data to be deleted is greater than
    # the root's data then it lies in right subtree
    elif data > root.data:
        root.right = delete_node(root.right, data)

    # if data is same as root's data, then this is the node to be deleted
    else:
        # if either the left or right child is null, return that node
        if root.left is None:
            return root.right
        elif root.right is None:
            return root.left

        # replace the root with the leftmost leaf of the right tree
        root.data = leftmost_leaf(root.right)

        # delete the duplicate leaf node
        root.right = delete_node(root.right, root.data)

    return root


if __name__ == "__main__":

    root = Node(4)
    root.left = Node(2)
    root.right = Node(6)
    root.left.left = Node(1)
    root.left.right = Node(3)
    root.right.left = Node(5)
    root.right.right = Node(7)

    # print tree before deletion
    print("Tree before root deletion: ")
    print_in_order(root)
    print("\n")

    # delete root node and print tree again
    delete_node(root, root.data)
    print("Tree after root deletion: ")
    print_in_order(root)
